### ISSUE TITLE
Bug: Envoy proxy blocks WebSockets

### ISSUE THREAD

**Issue by GrandathePanda** [aws/aws-app-mesh-roadmap#119]
*SECURITY NOTICE: If you think you’ve found a potential security issue, please do not post it in the Issues. Instead, please follow the instructions [here](https://aws.amazon.com/security/vulnerability-reporting/) or [email AWS security directly](mailto:aws-security@amazon.com).*

**Summary**
What are you observing that doesn't seem right?
With the envoy proxy websockets fail to connect to my elixir service.
Removing the envoy proxy allows websockets to connect like normal.
**Steps to Reproduce**
What are the steps you can take to reproduce this issue?
Add the app mesh envoy to an ecs service that can handle websockets.
**Are you currently working around this issue?**
How are you currently solving this problem?
Still working on that.
**Additional context**
Anything else we should know?
The envoy route configuration supports a boolean flag for upgrading websocket requests. I may be mistaken but it looks like that can't be configured from the ECS ui.
**Attachments**
If you think you might have additional information that you'd like to include via an attachment, please do - we'll take a look. (Remember to remove any personally-identifiable information.)

**Comment by dastbe** [comment]
Thanks for the bug report @GrandathePanda , and you are correct that we aren't allowing websocket on http1 routes today

We need to define how you would model [use_websocket](https://www.envoyproxy.io/docs/envoy/v1.5.0/api-v1/route_config/route#config-http-conn-man-route-table-route-use-websocket), and then :shipit:

**Comment by GrandathePanda** [comment]
So I know in my case it was somewhat unexpected that WebSockets weren't being upgraded but to be fair this is something I would have had to do manually in my config for other proxies as well I just overlooked it because the AWS UI tends to hide much of the usual complexity. This was just a case where that actually wound up not working.
 
To be clear this was from an NLB to the ECS service not from another mesh service.

With that in mind:

To question 1: What I'm getting at is that given you have to manually configure WS in other proxies as well I would not always have the WS requests upgraded but, I would have an opt-in much like the opt-in for app mesh itself on ECS. 

To question 2: I think it should be an http1 route with some degree of freedom in configuration

To question 3: If you mean a distinct listener in appmesh itself, then at least for me this wouldn't solve the issue it would have to be at the Envoy level since this was blocking WS requests through my NLB.

Those are my two cents, I'm sure there are many other angles and use cases to consider but I think I had a fairly standard usecase as It was just a web application pushing things to the react front end through the WS and using app mesh to more easily talk to some other containers in my ECS cluster.

**Comment by 0xMH** [comment]
We hit this issue lately as well... Took us some time to realize it's a problem with envoy sidecar itself.  Our use-case was close to OPs. We wanted to connect to our Appmesh service through ALB but the web socket connection didn't get through. Any workarounds for now?

**Comment by efe-selcuk** [comment]
We have an engineer looking to see what it would take to enable this

> Any workarounds for now?

From what I can tell, currently to handle websockets you'd have to use a TCP listener.

> also is it possible to pack our own sidecar image with use_websocket and use it?

Since the route configuration (among other things) is delivered dynamically to the proxy, you wouldn't be able to set this with a bootstrap configuration.

**Comment by unicomp21** [comment]
I suspect this won't be fixed for a long long time.  Maybe the better investment would be to start focusing on envoy http/3 streams?

https://quicwg.org/base-drafts/draft-ietf-quic-http.html#name-push-streams

**Comment by bcelenza** [comment]
Hey @unicomp21, I plan on looking into adding support for this next week actually. Do you have a need for WebSockets or Quic streams (or both?). We don't yet have a feature issue for Quic, so we can create that if needed.

**Comment by unicomp21** [comment]
No, I don't have a need.  I was just trying to point out that adding websockets will likely create a lot of confusion.  What happens when there are routing changes at the envoy level?  Do long running websockets get killed?  Are they allowed to linger?

**Comment by mquidi86** [comment]
I am migrating an app that I have already deployed to ECS, to use App Mesh. Everything was looking fine, I was creating the Cloudwatch dashboard when I noticed 400s response codes went up noticeably. It turns out it, plenty of 403s from the websockets. If there's no workaround I guess I'll have to rollback my changes for the moment.

**Comment by bcelenza** [comment]
Sorry for the delay in responding here. We're working on websocket support now, and plan to roll it out soon.

**Comment by abaptiste** [comment]
@unicomp21 My understanding is that draining occurs when the listener changes:

https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/operations/draining.html

That said, I will do some tests and update this issue once I have some concrete data on how routing changes affect long lived WebSocket connections.
